### PR TITLE
fix: fall back to an older release when the cool-down holds the latest

### PR DESCRIPTION
**Pull request by gsong**
### Description

Fixes the fallback half of #54. (The `--json` half of that issue is stacked in a follow-up PR.)

**Problem.** When an action's latest release was younger than `--min-age`, the run dropped the action instead of offering the newest release that does clear the cool-down. The `--mode` step-down added in v1.12.0 already behaved this way; the age filter in `cli/index.ts` ran ahead of it and removed the candidate outright. Two consequences:

- No fallback. `getCompatibleUpdate()` took a mode but no age constraint, so it could not express "newest release older than N days".
- Ordering. With `--mode minor --min-age 7`, an action whose latest is a too-new major was dropped for age before the mode logic could offer an eligible minor. And a version the mode substituted never had its own age checked.

**Solution.** Both constraints now go through `getCompatibleUpdate()`:

- `findCompatibleTag` becomes `findCompatibleTags`, returning every compatible tag ranked newest first, and accepts `major`. The default mode is `major`, so without this the fallback would never run for the reported case.
- `getCompatibleUpdate()` takes an optional `minAgeMs` (and `now`). It walks the ranked candidates and takes the first one old enough. Dates come from `getTagInfo`, one candidate at a time, so a one-release step-down does not pay for the whole tag history. Tag listings carry no dates, which is why this cannot be a single request.
- It returns `{ update, reason }` instead of a bare value. `reason` is `cool-down` when the mode had candidates but all were too young, and `no-candidate` when the mode allows nothing newer. The CLI uses that to put the action in the right notice: the cool-down owns the outcome whenever the mode still had something to offer.
- `cli/index.ts` decides mode and age in one pass. An action is reported as held back only when no release satisfies both.

The substituted update also carries its own `publishedAt` and a recomputed `isBreaking`, instead of the previous unconditional `isBreaking: false` and the stale date of the version that was rejected.

**Verified against the repro in the issue** with a build of this branch, a workflow pinned to `zizmorcore/zizmor-action@v0.6.0`:

```console
$ actions-up --dry-run --min-age 7
zizmorcore/zizmor-action: v0.6.0 → v0.6.2 (3dc1ecc)

$ actions-up --dry-run --min-age 0
zizmorcore/zizmor-action: v0.6.0 → v0.6.3 (70fb788)

$ actions-up --dry-run --min-age 90
⏳ Skipped 1 update released less than 90 days ago (cool-down, use --min-age 0 to disable)
   • zizmorcore/zizmor-action@v0.6.0
```

`pnpm test` passes in full, `test:packages` included. Coverage stays at 100%.

**Review follow-ups.** Three commits answer @azat-io's review:

- `fix: report a rate limited tag lookup instead of an undated tag`. `makeRequest` attaches `status: 403`, and the refs fallback in `get-tag-info.ts` checked for `status` before the rate-limit wrapper, so a rate-limited `getTagInfo` cached and returned `null` and both `@throws` contracts were dead. It now recognizes the rate limit first, the way `get-tag-sha.ts` already does.
- `fix: never step down to a prerelease or an unreleased tag`. The ranking skips prereleases unless the current version is one, mirroring `select-latest-family-tag.ts`, and `getCompatibleUpdate` now takes the vetted `latestVersion` as a ceiling.
- `perf: look a blocked action up once for all of its occurrences`. The lookup is memoized per action and version with a promise-valued map, so N workflows using one blocked action no longer fire N tag listings and date walks.

### Reviewer notes

- The age-vs-mode attribution rule is in `cli/index.ts`, in the branch that reads `compatible.reason`. Please check it reads the way you want in the `--mode minor --min-age N` combinations.
- Two tests in `test/cli/index.test.ts` changed fixture rather than expectation. They used a too-new **major** bump to exercise the min-age notice under `--mode minor`. Under the corrected logic that case is genuinely mode-blocked (no eligible minor exists at all), so the fixtures became too-new **minor** bumps.
- `core/versions/find-compatible-tag.ts` is renamed to the plural. The old singular had one caller, and keeping a wrapper would have tripped `knip`.
- `cli/build-json-report.ts` is untouched here; it moves in the follow-up.

---

### Pre-merge checklist

- [x] No similar open PR exists
- [x] Description explains problem/solution or links an issue
- [x] Tests added/updated when needed

### CODE REVIEW

File: core/api/get-compatible-update.ts [hunk 1]
```diff
@@ -76,32 +154,127 @@ export async function getCompatibleUpdate(
           await client.getAllTags(owner, repo, 100)
         : await client.getMatchingTagReferences(owner, repo, familyPrefix)
     } catch {
-      return null
+      return { reason: 'no-candidate', update: null }
```
[the remark this answers is not in the document]
Comment (gsong) [reply core/api/get-compatible-update.ts:167]: Fair point about the contract, and I have corrected it in 090e6b4. `no-candidate` now documents what it actually covers:

> `no-candidate` means no compatible tag was established: either the listing held nothing newer that the mode allows, or the listing could not be read.

I did not add a distinct lookup-failure result, for two reasons:

1. The `catch` is unchanged from `main` — `getCompatibleUpdate` already returned `null` on a failed tag fetch, and the CLI already put that action in `blockedByMode`. This PR did not introduce the conflation, only a `reason` field whose wording overstated it.
2. A third reason needs a fourth bucket to land in. The CLI has `skipped`, `blockedByMode` and `blockedByAge`, and none of them means "we could not check". Adding one means a new notice and a new `--json` field, which is a bigger change than #54 asks for.

Two related things I would rather raise than smuggle in here, @azat-io — happy to open issues, or to add either to this PR if you prefer:

- That `catch` swallows `GitHubRateLimitError` too. Everywhere else (`check-updates.ts`, `resolveTagMeta`) rethrows it so `run()` can print the rate-limit notice. This matters more after this PR than before: under the default `major` mode, `main` never fetched tags at all, and now the cool-down step-down does.
- A failed tag fetch is reported to the user as a mode or cool-down holdback, which asserts something the run never verified.

File: cli/index.ts [hunk 2]
```diff
@@ -321,57 +310,74 @@ async function runUpdate(options: CLIOptions): Promise<void> {
...
-          mode === 'minor' ?
+        let allowedByMode =
+          mode === 'major' ||
+          (mode === 'minor' ?
             ['minor', 'patch', 'none']
-          : ['patch', 'none']).includes(level)
+          : ['patch', 'none']
+          ).includes(level)
+        let allowedByAge =
+          !update.publishedAt || now - update.publishedAt.getTime() >= minAgeMs
```
[the remark this answers is not in the document]
Comment (gsong) [reply cli/index.ts:329]: I am leaving this as is, because it is `main`'s behavior rather than something this PR introduces. The filter it replaces reads:

```ts
outdated = outdated.filter(update => {
  if (!update.publishedAt) {
    return true
  }
  ...
})
```

So "an unknown publication date does not block an update" is the existing contract, and `allowedByAge` preserves it exactly. Flipping it to fail closed would hold back every action whose date cannot be resolved — including on a rate-limited run — which is a behavior change well beyond the fallback #54 asks for. @azat-io, that is your call to make rather than mine; say the word and I will add it here or open a separate issue.

One correction to the finding: line 320 reads `update.publishedAt`, which `checkUpdates` produces. `resolveTagMeta` is not involved, and this branch does not bypass `getCompatibleUpdate` — it is reached only when the mode is also satisfied, in which case there is nothing to step down to.

There is a narrower version of your point that is genuinely new here, and I want to flag it rather than let it pass. `selectAgedCandidate` in `core/api/get-compatible-update.ts` also treats an undated candidate as old enough. So if the latest release is held by the cool-down and the next candidate down has no resolvable date, that candidate is offered without being verified against `--min-age`. It can never be newer than the release that was just rejected, since candidates are walked newest first, but it could still sit inside the cool-down window. I chose consistency with the top-level rule and documented it in the JSDoc. If you would rather the step-down fail closed while the top-level check keeps failing open, that is a small change and I am glad to make it.

File: core/api/get-compatible-update.ts [hunk 3]
```diff
@@ -76,32 +155,127 @@ export async function getCompatibleUpdate(
...
+        return selected
+      }
+
+      let meta = await resolveTagMeta(client, {
+        tag: candidate.tag,
+        owner,
+        repo,
+      })
+
+      if (meta.date && now - meta.date.getTime() < minAgeMs) {
```
[the remark this answers is not in the document]
Comment (gsong) [reply core/api/get-compatible-update.ts:264]: This is the point I raised myself in the thread on `cli/index.ts:320`, so I will not re-answer the policy half here: whether an undated candidate should stay ineligible while the top-level check keeps failing open is @azat-io's call, and I have offered to make that change either way.

There is a part of it I can confirm is a defect this PR introduced, and I have fixed it in b47afb6.

The `resolveTagMeta` this PR adds at `core/api/get-compatible-update.ts:270` swallowed every error. Its twin at `core/api/check-updates.ts:769` rethrows `GitHubRateLimitError`. So on a rate-limited run the new copy returned `date: null` for every candidate, `selectAgedCandidate` took the newest one as old enough, and the run offered a release inside the cool-down while saying nothing about the rate limit. It now rethrows, and `run()` already catches that and prints the rate-limit notice. Test added; coverage stays at 100%.

That is the mechanism by which "metadata cannot be read" was reachable in practice. What remains is a tag whose date GitHub genuinely does not carry, which `main` has always treated as old enough — a contract, not an oversight, and the one still awaiting @azat-io's answer.

One correction to the finding as written: `resolveTagMeta` does not return `date: null` on a rejection any more, and the "returns `null`" case is `getTagInfo` resolving to `null`, which is a different path from a rejection.

File: core/versions/find-compatible-tags.ts [hunk 4]
```diff
@@ -47,7 +52,7 @@ export function findCompatibleTag(
       continue
     }
 
-    if (semver.major(family.version) !== currentMajor) {
+    if (mode !== 'major' && semver.major(family.version) !== currentMajor) {
```
Comment (azat-io) [inline core/versions/find-compatible-tags.ts:75]: The walk can offer prereleases: `parseTagFamily` keeps the qualifier and `semver.gt` passes them, and `rcompare` ranks `v4.1.0-rc.1` right below `v4.1.0`. An RC is usually published days before its stable, so a too-young stable often steps down to its own RC — and with the defaults (`--min-age 1`, mode `major`) this is reachable for any action whose latest release is under a day old. The walk also isn't capped at the release-vetted `latestVersion`, so it can land on an unreleased tag.

Please skip candidates with a prerelease qualifier unless the current version has one (mirroring `select-latest-family-tag.ts`), and consider capping the walk at `latestVersion`.

File: core/api/get-compatible-update.ts [hunk 5]
```diff
@@ -76,32 +157,135 @@ export async function getCompatibleUpdate(
...
+async function resolveTagMeta(
+  client: GitHubClient,
+  parameters: { owner: string; repo: string; tag: string },
+): Promise<{ sha: string | null; date: Date | null }> {
+  try {
+    let { owner, repo, tag } = parameters
+    let info = await client.getTagInfo(owner, repo, tag)
+    return { date: info?.date ?? null, sha: info?.sha ?? null }
+  } catch (error) {
+    if (error instanceof Error && error.name === 'GitHubRateLimitError') {
```
Comment (azat-io) [inline core/api/get-compatible-update.ts:296]: This rethrow can't fire for a real rate limit: `makeRequest` attaches `status: 403`, and `get-tag-info.ts` checks `'status' in tagError` before its rate-limit wrapper, so production `getTagInfo` returns (and caches) `null` instead of throwing. The unit test passes only because it injects `GitHubRateLimitError` directly — a shape production never produces.

To be fair, `main` behaves the same today (`check-updates.ts` has the identical dead rethrow), so this isn't a regression — but the commit message and the `@throws` contracts promise a guarantee that doesn't hold. Checking `error.message.includes('rate limit')` before the status swallow in `get-tag-info.ts` (the way `get-tag-sha.ts` already does) would make both rethrows real.

File: cli/index.ts [hunk 6]
```diff
@@ -284,31 +284,20 @@ async function runUpdate(options: CLIOptions): Promise<void> {
...
 
-    let blockedByMode: typeof outdated = []
-    if (mode !== 'major') {
-      let tagsCache = new Map<
-        string,
-        Awaited<ReturnType<typeof githubClient.getAllTags>>
-      >()
-      let shaCache = new Map<string, string | null>()
-      let decisions = outdated.map(update => {
+    let decisions = await Promise.all(
```
Comment (azat-io) [inline cli/index.ts:308]: This maps per-occurrence updates, and `tagsCache`/`shaCache` store values, not promises — so N workflows using the same blocked action fire N concurrent listings and date walks (each uncached `getTagInfo` is up to ~3 requests). On the unauthenticated 60 req/h budget the duplication alone can trigger the rate-limit abort. The resolution phase below already dedupes with a promise-valued memo (`resolutionCache`); doing the same per `action@version` here would fix it.

File: pnpm-lock.yaml [hunk 7]
```diff
@@ -25,8 +25,6 @@ settings:
   autoInstallPeers: true
   excludeLinksFromLockfile: false
 
-pnpmfileChecksum: sha256-15yQafiOk/n/FnyMLJdjllwGXgNxN7Ro+mGuxVKxhic=
```
Comment (azat-io) [inline pnpm-lock.yaml:28]: This hunk looks like an environment artifact — the checksum comes from the `configDependencies` plugin in `pnpm-workspace.yaml`, and CI installs with `--frozen-lockfile` under pnpm 11 where it's expected. Please drop it from the PR.

**Review by azat-io — COMMENTED** [review]
Thanks — the direction is right, and the core of the step-down (the ranked walk, the age/mode attribution, the recomputed `isBreaking`) looks solid. A few things I'd like addressed before merge, inline below.

File: core/versions/find-compatible-tags.ts (hunk 4 above)
Comment (gsong) [reply core/versions/find-compatible-tags.ts:75]: Fixed in 5c4550a. `findCompatibleTags` now skips prerelease candidates unless the current version is itself a prerelease, mirroring `selectLatestFamilyTag`, and I took the cap as well rather than leaving it as a maybe.

The cap goes through `getCompatibleUpdate`, which now takes the `latestVersion` the update check already settled on and passes it down as a ceiling: a tag above it is dropped before ranking. That closes the unreleased-tag hole you named. A `latestVersion` that carries no comparable core (a branch, a SHA) parses to null and imposes no ceiling, which keeps the existing "no opinion" contract of `parseTagFamily`.

Your reachability note was the part I had not seen. With the defaults the step-down runs whenever the latest release is under a day old, so an RC published days before its stable was a likely landing spot rather than a corner case.

Tests added for both guards; coverage stays at 100%. The readme paragraph now states the two limits.

File: core/api/get-compatible-update.ts (hunk 5 above)
Comment (gsong) [reply core/api/get-compatible-update.ts:296]: You are right, and I confirmed it: `makeRequest` sets `error.status = 403` before it rewrites the message, and the refs fallback in `get-tag-info.ts` tested `'status' in tagError` first, so production `getTagInfo` cached and returned `null` on a rate limit. The wrapper at the end of the function could never fire. My test injected the error shape directly and proved nothing.

Fixed in 90ca4b7, the way you suggested: `get-tag-info.ts` now checks `error.message.includes('rate limit')` before the status swallow, matching `get-tag-sha.ts`. Both rethrows are now real, so `run()` prints the rate-limit notice instead of the run reading a rate-limited lookup as a tag with no publication date. Added a test that drives it through a real 403 response with a rate-limit body, rather than through an injected error.

This does change `main` behavior beyond #54, since `check-updates.ts` shares the same producer. I think that is the correction rather than a scope creep, but flag it if you would rather it landed on its own.

File: cli/index.ts (hunk 6 above)
Comment (gsong) [reply cli/index.ts:308]: Fixed in ec6fe95, with the promise-valued memo you pointed at. The lookup is now keyed by action name, effective current version and latest version, so N occurrences of one blocked action share a single tag listing and date walk. Test added that asserts one call for two occurrences.

The memo is populated before the first `await`, so occurrences that start together join the same promise rather than racing past the check.

File: pnpm-lock.yaml (hunk 7 above)
Comment (gsong) [reply pnpm-lock.yaml:28]: Already gone. That comment anchors to b47afb6, which I force-pushed away before you reviewed; the PR has not touched `pnpm-lock.yaml` since. `gh api .../pulls/55/files` lists seven files and the lockfile is not among them.

You diagnosed it correctly, though: local `pnpm` runs keep rewriting that `pnpmfileChecksum` line, so I reverted it again while preparing these commits. `pnpm test` now passes in full on this branch, `test:packages` included.

File: core/api/get-compatible-update.ts [hunk 8]
```diff
@@ -76,32 +164,138 @@ export async function getCompatibleUpdate(
...
     tagsCache.set(tagsCacheKey, tags)
   }
 
-  let compatibleTag = findCompatibleTag(tags, currentVersion, mode)
-  if (!compatibleTag) {
-    return null
+  let candidates = findCompatibleTags(tags, currentVersion, {
+    latestVersion: parameters.latestVersion,
+    mode,
+  })
```
[the remark this answers is not in the document]
Comment (gsong) [reply core/api/get-compatible-update.ts:175]: I checked this against how `latestVersion` is produced, and I am not making the change: it would be stricter than the rule that set the ceiling.

`check-updates.ts` does not require a release. The family path takes `selectLatestFamilyTag(familyTags, …)` at `core/api/check-updates.ts:256`, which reads the tag listing. The release path falls through to `client.getAllTags(…)` at lines 344 and 439 whenever there is no release, the release tag is a moving major, or its version does not parse. So for an action that publishes git tags and no GitHub releases, `latestVersion` is itself a plain tag.

Rejecting candidates without a release would therefore switch the step-down off entirely for those repositories. The ceiling would be a tag, and every candidate under it would be rejected for the same property that let the ceiling through. That is the case #54 asks to fix, so the change would regress it rather than harden it.

Two corrections to the finding as written:

- The cool-down still measures a real age for a tag with no release. `getTagInfo` falls back to the annotated tag's tagger date, then to the commit author date. That is the date the tag came into existence, which is what `--min-age` asks about. An unreleased candidate is age-checked, not waved through.
- `main` already selects from the tag listing without checking for a release; the `--mode` step-down added in v1.12.0 works that way. This PR narrowed it with the `latestVersion` ceiling. Requiring a release would be a further behavior change beyond #54.

@azat-io, the ceiling you asked for closes the unreleased-tag hole above `latestVersion`. If you also want candidates below it to be release-backed, that is a policy call and it needs the tag-only repositories answered first. Say the word and I will open an issue.

File: test/api/get-tag-info.test.ts [hunk 9]
```diff
@@ -881,6 +881,25 @@ describe('getTagInfo', () => {
...
+        statusText: 'Forbidden',
+        status: 403,
+      })
+      return Promise.resolve(response)
+    })
+
+    await expect(
+      getTagInfo(context, { tag: 'v5.1.0', owner: 'o', repo: 'r' }),
+    ).rejects.toHaveProperty('name', 'GitHubRateLimitError')
+    expect(context.caches.tagInfo.has('o/r#v5.1.0')).toBeFalsy()
```
[the remark this answers is not in the document]
Comment (gsong) [reply test/api/get-tag-info.test.ts:900]: Confirmed, and fixed in 0340bf4.

You are right about the reachability. `getTagInfo` recognized the rate limit only in the refs fallback, at the end of the chain. The nested annotated-tag and commit requests each had a best-effort catch that swallowed everything. So release 404 → reference 200 → commit 403 cached a `TagInfo` with `date: null`, and the cool-down read that as a tag old enough to offer while the run said nothing about the rate limit.

Every catch in the chain now rethrows a recognized rate limit: both nested paths in the release branch, both in the refs fallback, and the reference request itself. The release branch also rethrows before retrying the refs fallback, so a rate-limited run no longer spends a second request on an exhausted budget.

Tests added for all four nested paths, driven through a real 403 with a rate-limit body rather than an injected error, plus one for a non-rate-limit commit failure so the best-effort path stays covered. `pnpm test` passes in full and coverage stays at 100%.

One thing I did not change. `get-tag-sha.ts` has the same swallow in its annotated-tag lookup: on a rate limit it answers with the tag object SHA instead of the commit SHA, and caches it. Same class of defect, different file, and this PR does not reach it. @azat-io, tell me whether you want that here or in its own issue.

**Comment by azat-io** [comment]
@gsong Since #56's squash (571c993) brought this whole branch into main, I think this PR can simply be closed — nothing left to merge. Am I missing anything on your side? I'll get to the open threads separately.

**Comment by gsong** [comment]
> @gsong Since #56's squash ([571c993](https://github.com/azat-io/actions-up/commit/571c9933f7f0231c195f5ee110970bc5670e9f06)) brought this whole branch into main, I think this PR can simply be closed — nothing left to merge. Am I missing anything on your side? I'll get to the open threads separately.

Yes, #56 is a superset of this PR.

**Comment by gsong** [comment]
Already landed in main via the squash merge of #56, which carried this branch's commits. Closing as merged-in-substance.
